File: sdk/go/interceptors.go
package micropod

import (
	"context"
	"math/rand/v2"
	"time"

	"connectrpc.com/connect"
	"connectrpc.com/otelconnect"
	"go.opentelemetry.io/otel"
	"go.opentelemetry.io/otel/metric"
	"go.opentelemetry.io/otel/trace"
)

// RetryPolicy controls unary-call retry behavior.
type RetryPolicy struct {
	// MaxAttempts is the total number of tries including the first (default 3).
	MaxAttempts int
	// InitialBackoff before the first retry (default 100ms).
	InitialBackoff time.Duration
	// MaxBackoff caps the exponential growth (default 2s).
	MaxBackoff time.Duration
	// Multiplier applied per retry (default 2.0).
	Multiplier float64
	// RetryableCodes — Connect codes worth retrying (default: unavailable,
	// deadline_exceeded, resource_exhausted, aborted).
	RetryableCodes []connect.Code
}

// DefaultRetryPolicy: 3 attempts, 100ms → ~2s backoff with ±25% jitter.
func DefaultRetryPolicy() RetryPolicy {
	return RetryPolicy{
		MaxAttempts:    3,
		InitialBackoff: 100 * time.Millisecond,
		MaxBackoff:     2 * time.Second,
		Multiplier:     2.0,
		RetryableCodes: []connect.Code{
			connect.CodeUnavailable,
			connect.CodeDeadlineExceeded,
			connect.CodeResourceExhausted,
			connect.CodeAborted,
		},
	}
}

func (p RetryPolicy) retryable(code connect.Code) bool {
	for _, c := range p.RetryableCodes {
		if c == code {
			return true
		}
	}
	return false
}

// retryInterceptor retries unary calls on transient Connect codes with
// exponential backoff + jitter. Unary request bodies are buffered by
// connect-go, so replaying next() is safe.
type retryInterceptor struct {
	policy RetryPolicy
}

func (r retryInterceptor) WrapUnary(next connect.UnaryFunc) connect.UnaryFunc {
	return func(ctx context.Context, req connect.AnyRequest) (connect.AnyResponse, error) {
		attempts := r.policy.MaxAttempts
		if attempts < 1 {
			attempts = 1
		}
		backoff := r.policy.InitialBackoff
		var lastErr error
		for i := 0; i < attempts; i++ {
			resp, err := next(ctx, req)
			if err == nil {
				return resp, nil
			}
			lastErr = err
			code := connect.CodeOf(err)
			if !r.policy.retryable(code) || i == attempts-1 {
				return nil, err
			}
			// ±25% jitter
			jitter := time.Duration(rand.Int64N(int64(backoff/2) + 1))
			delay := backoff - backoff/4 + jitter
			timer := time.NewTimer(delay)
			select {
			case <-ctx.Done():
				timer.Stop()
				return nil, ctx.Err()
			case <-timer.C:
			}
			backoff = time.Duration(float64(backoff) * r.policy.Multiplier)
			if backoff > r.policy.MaxBackoff {
				backoff = r.policy.MaxBackoff
			}
		}
		return nil, lastErr
	}
}

func (r retryInterceptor) WrapStreamingClient(next connect.StreamingClientFunc) connect.StreamingClientFunc {
	// Streams are not retried — partial progress can't be replayed safely.
	return next
}

func (r retryInterceptor) WrapStreamingHandler(next connect.StreamingHandlerFunc) connect.StreamingHandlerFunc {
	return next
}

// timeoutInterceptor applies a default deadline when the incoming context
// has none or a later one.
type timeoutInterceptor struct {
	timeout time.Duration
}

func (t timeoutInterceptor) WrapUnary(next connect.UnaryFunc) connect.UnaryFunc {
	return func(ctx context.Context, req connect.AnyRequest) (connect.AnyResponse, error) {
		if t.timeout > 0 {
			if deadline, ok := ctx.Deadline(); !ok || time.Until(deadline) > t.timeout {
				var cancel context.CancelFunc
				ctx, cancel = context.WithTimeout(ctx, t.timeout)
				defer cancel()
			}
		}
		return next(ctx, req)
	}
}

func (t timeoutInterceptor) WrapStreamingClient(next connect.StreamingClientFunc) connect.StreamingClientFunc {
	return func(ctx context.Context, spec connect.Spec) connect.StreamingClientConn {
		if t.timeout > 0 {
			if deadline, ok := ctx.Deadline(); !ok || time.Until(deadline) > t.timeout {
				var cancel context.CancelFunc
				ctx, cancel = context.WithTimeout(ctx, t.timeout)
				return &cancelOnCloseConn{StreamingClientConn: next(ctx, spec), cancel: cancel}
			}
		}
		return next(ctx, spec)
	}
}

func (t timeoutInterceptor) WrapStreamingHandler(next connect.StreamingHandlerFunc) connect.StreamingHandlerFunc {
	return next
}

type cancelOnCloseConn struct {
	connect.StreamingClientConn
	cancel context.CancelFunc
}

func (c *cancelOnCloseConn) CloseRequest() error {
	err := c.StreamingClientConn.CloseRequest()
	c.cancel()
	return err
}

// OTelOption customizes the OpenTelemetry interceptor.
type OTelOption func(*otelConfig)

type otelConfig struct {
	tracerProvider trace.TracerProvider
	meterProvider  metric.MeterProvider
}

// WithTracerProvider sets the OTel tracer provider (default: global).
func WithTracerProvider(tp trace.TracerProvider) OTelOption {
	return func(c *otelConfig) { c.tracerProvider = tp }
}

// WithMeterProvider sets the OTel meter provider (default: global).
func WithMeterProvider(mp metric.MeterProvider) OTelOption {
	return func(c *otelConfig) { c.meterProvider = mp }
}

// newOTelInterceptor builds the otelconnect interceptor — spans per call
// with rpc.system/rpc.service/rpc.method attributes, duration/request
// metrics, and W3C traceparent propagation into request headers.
func newOTelInterceptor(opts ...OTelOption) connect.Interceptor {
	cfg := &otelConfig{
		tracerProvider: otel.GetTracerProvider(),
		meterProvider:  otel.GetMeterProvider(),
	}
	for _, o := range opts {
		o(cfg)
	}
	otelOpts := []otelconnect.Option{
		otelconnect.WithTracerProvider(cfg.tracerProvider),
		otelconnect.WithMeterProvider(cfg.meterProvider),
		otelconnect.WithPropagator(otel.GetTextMapPropagator()),
	}
	i, err := otelconnect.NewInterceptor(otelOpts...)
	if err != nil {
		// NewInterceptor only errors on nil options misuse; degrade to a
		// no-op rather than panic a client constructor.
		return connect.UnaryInterceptorFunc(func(next connect.UnaryFunc) connect.UnaryFunc {
			return next
		})
	}
	return i
}

File: sdk/ts/src/interceptors.ts
import { Code, ConnectError, type Interceptor } from "@connectrpc/connect";
import { createValidator } from "@bufbuild/protovalidate";
import {
  context,
  metrics,
  trace,
  SpanKind,
  SpanStatusCode,
  type Span,
  type Tracer,
  type Meter,
} from "@opentelemetry/api";

// ---------------------------------------------------------------------------
// Retry — unary only, transient codes, exponential backoff + jitter.
// ---------------------------------------------------------------------------

export interface RetryPolicy {
  /** Total tries including the first (default 3). */
  maxAttempts?: number;
  /** Backoff before the first retry (default 100ms). */
  initialBackoffMs?: number;
  /** Backoff cap (default 2000ms). */
  maxBackoffMs?: number;
  /** Growth per retry (default 2). */
  multiplier?: number;
  /** Connect codes worth retrying. */
  retryableCodes?: Code[];
}

export const defaultRetryPolicy: Required<RetryPolicy> = {
  maxAttempts: 3,
  initialBackoffMs: 100,
  maxBackoffMs: 2000,
  multiplier: 2,
  retryableCodes: [
    Code.Unavailable,
    Code.DeadlineExceeded,
    Code.ResourceExhausted,
    Code.Aborted,
  ],
};

function sleep(ms: number, signal: AbortSignal | undefined): Promise<void> {
  return new Promise((resolve, reject) => {
    const t = setTimeout(() => {
      signal?.removeEventListener("abort", onAbort);
      resolve();
    }, ms);
    const onAbort = () => {
      clearTimeout(t);
      reject(ConnectError.from(signal?.reason ?? new Error("aborted")));
    };
    signal?.addEventListener("abort", onAbort, { once: true });
  });
}

export function retryInterceptor(policy: RetryPolicy = {}): Interceptor {
  const p = { ...defaultRetryPolicy, ...policy };
  return (next) => async (req) => {
    if (req.stream) {
      // Streams can't be replayed mid-flight — no retry.
      return next(req);
    }
    let backoff = p.initialBackoffMs;
    let attempt = 0;
    for (;;) {
      attempt++;
      try {
        return await next(req);
      } catch (e) {
        const err = ConnectError.from(e);
        if (attempt >= p.maxAttempts || !p.retryableCodes.includes(err.code)) {
          throw err;
        }
        const jitter = Math.random() * (backoff / 2);
        await sleep(backoff - backoff / 4 + jitter, req.signal);
        backoff = Math.min(backoff * p.multiplier, p.maxBackoffMs);
      }
    }
  };
}

// ---------------------------------------------------------------------------
// Timeout — default per-call deadline when the caller set none (or a later
// one). Implemented via signal composition so it works on every runtime.
// ---------------------------------------------------------------------------

export function timeoutInterceptor(timeoutMs: number): Interceptor {
  return (next) => async (req) => {
    if (timeoutMs <= 0) return next(req);
    const ctrl = new AbortController();
    const timer = setTimeout(
      () => ctrl.abort(new ConnectError("deadline exceeded", Code.DeadlineExceeded)),
      timeoutMs,
    );
    const onAbort = () => ctrl.abort(req.signal?.reason);
    req.signal?.addEventListener("abort", onAbort, { once: true });
    try {
      return await next({ ...req, signal: ctrl.signal });
    } finally {
      clearTimeout(timer);
      req.signal?.removeEventListener("abort", onAbort);
    }
  };
}

// ---------------------------------------------------------------------------
// Validation — evaluates the buf.validate constraints declared in the protos
// against each request before it hits the wire; invalid requests fail fast
// with invalid_argument instead of a server round-trip.
// ---------------------------------------------------------------------------

export function validationInterceptor(): Interceptor {
  const validator = createValidator();
  return (next) => async (req) => {
    // Client/bidi streams send an AsyncIterable of messages — the micropod.v1
    // services have none, but guard so the interceptor stays safe elsewhere.
    if (req.stream) return next(req);
    const result = validator.validate(req.method.input, req.message);
    if (result.kind === "invalid") {
      const detail = result.violations
        .map((v) => `${String(v.field)}: ${v.message}`)
        .join("; ");
      throw new ConnectError(`invalid request — ${detail}`, Code.InvalidArgument);
    }
    return next(req);
  };
}

// ---------------------------------------------------------------------------
// OpenTelemetry — a span per call with rpc.* attributes, W3C traceparent
// propagation, duration histogram + call counter. Uses the global OTel
// providers by default; pass a tracer/meter to override.
// ---------------------------------------------------------------------------

export interface OtelOptions {
  tracer?: Tracer;
  meter?: Meter;
}

const RPC_SYSTEM = "connect_rpc";

function injectTraceparent(header: Headers, span: Span): void {
  const sc = span.spanContext();
  if (!sc.traceId || !sc.spanId) return;
  const flags = (sc.traceFlags & 1) === 1 ? "01" : "00";
  header.set("traceparent", `00-${sc.traceId}-${sc.spanId}-${flags}`);
}

export function otelInterceptor(opts: OtelOptions = {}): Interceptor {
  const tracer = opts.tracer ?? trace.getTracer("@micropod/sdk");
  const meter = opts.meter ?? metrics.getMeter("@micropod/sdk");
  const duration = meter.createHistogram("micropod.client.duration", {
    unit: "ms",
    description: "Client RPC duration",
  });
  const calls = meter.createCounter("micropod.client.calls", {
    description: "Client RPC calls",
  });

  const finish = (
    span: Span,
    service: string,
    method: string,
    start: number,
    code: Code | null,
  ) => {
    const attrs = {
      "rpc.system": RPC_SYSTEM,
      "rpc.service": service,
      "rpc.method": method,
      ...(code !== null ? { "rpc.connect.code": Code[code] } : {}),
    };
    calls.add(1, attrs);
    duration.record(Date.now() - start, attrs);
    span.setAttributes(attrs);
    span.setStatus(
      code === null
        ? { code: SpanStatusCode.OK }
        : { code: SpanStatusCode.ERROR, message: Code[code] },
    );
    span.end();
  };

  return (next) => async (req) => {
    const service = req.service.typeName;
    const method = req.method.name;
    const span = tracer.startSpan(
      `${service}/${method}`,
      { kind: SpanKind.CLIENT },
      context.active(),
    );
    injectTraceparent(req.header, span);
    const start = Date.now();
    try {
      const res = await context.with(trace.setSpan(context.active(), span), () =>
        next(req),
      );
      if (req.stream) {
        // Keep the span open until the stream finishes.
        const stream = (res as { message?: AsyncIterable<unknown> }).message;
        if (stream) {
          (res as { message: AsyncIterable<unknown> }).message = wrapStream(
            stream,
            () => finish(span, service, method, start, null),
            (e) => finish(span, service, method, start, ConnectError.from(e).code),
          );
        } else {
          finish(span, service, method, start, null);
        }
      } else {
        finish(span, service, method, start, null);
      }
      return res;
    } catch (e) {
      const err = ConnectError.from(e);
      span.recordException(err);
      finish(span, service, method, start, err.code);
      throw err;
    }
  };
}

async function* wrapStream(
  stream: AsyncIterable<unknown>,
  onEnd: () => void,
  onError: (e: unknown) => void,
): AsyncIterable<unknown> {
  try {
    for await (const m of stream) yield m;
    onEnd();
  } catch (e) {
    onError(e);
    throw e;
  }
}
